com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchWithoutDebuggingRequestHandler.java

```diff
@@ -0,0 +1,169 @@
+/*******************************************************************************
+* Copyright (c) 2018 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.debug.core.adapter.handler;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.logging.Logger;
+
+import com.google.gson.JsonObject;
+import com.microsoft.java.debug.core.Configuration;
+import com.microsoft.java.debug.core.DebugException;
+import com.microsoft.java.debug.core.adapter.ErrorCode;
+import com.microsoft.java.debug.core.adapter.IDebugAdapterContext;
+import com.microsoft.java.debug.core.adapter.ProcessConsole;
+import com.microsoft.java.debug.core.protocol.Events;
+import com.microsoft.java.debug.core.protocol.JsonUtils;
+import com.microsoft.java.debug.core.protocol.Messages.Request;
+import com.microsoft.java.debug.core.protocol.Messages.Response;
+import com.microsoft.java.debug.core.protocol.Requests.CONSOLE;
+import com.microsoft.java.debug.core.protocol.Requests.Command;
+import com.microsoft.java.debug.core.protocol.Requests.LaunchArguments;
+import com.microsoft.java.debug.core.protocol.Requests.RunInTerminalRequestArguments;
+import com.sun.jdi.connect.IllegalConnectorArgumentsException;
+import com.sun.jdi.connect.VMStartException;
+
+public class LaunchWithoutDebuggingRequestHandler implements ILaunchDelegateHandler {
+    protected static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+    protected static final String TERMINAL_TITLE = "Java Process Console";
+    protected static final long RUNINTERMINAL_TIMEOUT = 10 * 1000;
+
+    private Process launchInternalDebuggeeProcess(LaunchArguments launchArguments, IDebugAdapterContext context)
+            throws IOException, IllegalConnectorArgumentsException, VMStartException {
+        String[] cmds = CommonLaunchRequestHandler.constructLaunchCommands(launchArguments, false, null);
```

Comment:
why not make CommonLaunchRequestHandler  abstract class: AbstractLaunchRequestHandler  and extends it?

Reply:
Previous way was using the abstract class you mentioned, and there would be 2 derived LaunchRequestHandler for debug/noDebug mode respectively. If so, we have to register both handlers to respond to `launch` requests. (because we never know the launching mode before we handle the request)
By following delegation pattern, `launch` request is handled by the CommonLaunchRequestHandler, according to the noDebug argument, it then calls interfaces in different delegates. By putting those abstract methods to an interface, it's also clearer and easier to read.